### PR TITLE
Remote Environment Broker - Register namespaced broker on EnvironmentMapping creation

### PR DESCRIPTION
**Pull request by aszecowka**
Confirm these statements before you submit your pull request:

- [x] I have read and submitted the required [Contributor Licence Agreements](https://github.com/kyma-project/community/blob/master/CONTRIBUTING.md#agreements-and-licenses).
- [x] This pull request follows the contributing guidelines, recommended Git workflow, and templates.
- [x] I have tested my changes.
- [ ] I have updated the relevant documentation.
---
If REB is run in mode namespaced broker (ClusterScopedBrokerEnabled=false), we watch for the creation of EnvironmentMapping and create ServiceBroker and Service.

To test it on kyma, you need to change REB deployment:
```
        - name: APP_CLUSTER_SCOPED_BROKER_NAME
          value: core-remote-environment-broker
        - name: APP_CLUSTER_SCOPED_BROKER_ENABLED
          value: "false"
        - name: APP_UNIQUE_SELECTOR_LABEL_KEY
          value: app
        - name: APP_UNIQUE_SELECTOR_LABEL_VALUE
          value: core-remote-environment-broker
        - name: APP_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
```

And edit clusterorle:
```
- apiGroups:
  - servicecatalog.k8s.io
  resources:
  - servicebrokers
  verbs:
  - get
  - create
  - delete
- apiGroups:
  - ""
  resources:
  - services
  verbs:
  - create
  - delete
```

### CODE REVIEW

File: components/remote-environment-broker/internal/mapping/controller.go [hunk 1]
```diff
@@ -139,7 +152,7 @@ func (c *Controller) processNextItem() bool {
 	case err == nil:
 		c.queue.Forget(key)
 
-	case isTemporaryError(err) && c.queue.NumRequeues(key) < maxEnvironmentMappingProcessRetries:
+	case c.queue.NumRequeues(key) < maxEnvironmentMappingProcessRetries:
```
Comment (aszecowka) [inline components/remote-environment-broker/internal/mapping/controller.go]: handling temporary errors introduce a lot of difficulties - so we decided to simplify it and on every error retries can be performed.

File: components/remote-environment-broker/internal/mapping/controller.go [hunk 2]
```diff
@@ -37,25 +37,39 @@ type nsPatcher interface {
...
+	nsInformer      cache.SharedIndexInformer
+	nsPatcher       nsPatcher
+	reGetter        reGetter
+	nsBrokerFacade  nsBrokerFacade
+	log             logrus.FieldLogger
 }
 
 // New creates new environment mapping controller
-func New(emInformer cache.SharedIndexInformer, nsInformer cache.SharedIndexInformer, nsPatcher nsPatcher, reGetter reGetter, log logrus.FieldLogger) *Controller {
+func New(manageNsBrokers bool, systemNamespace string, emInformer cache.SharedIndexInformer, nsInformer cache.SharedIndexInformer, nsPatcher nsPatcher, reGetter reGetter, nsBrokerFacade nsBrokerFacade, log logrus.FieldLogger) *Controller {
```
Comment (mszostok) [inline components/remote-environment-broker/internal/mapping/controller.go]: suggestion: IMO `systemNamespace` can be called `workingNamespace`

File: components/remote-environment-broker/internal/mapping/controller.go [hunk 3]
```diff
@@ -157,47 +171,96 @@ func (c *Controller) processItem(key string) error {
...
+		}
+		return nil
+	}
+	if err := c.ensureNsLabelled(name, reNs); err != nil {
+		return err
+	}
+	if c.manageNsBrokers {
+		envMapping, ok := emObj.(*v1alpha1.EnvironmentMapping)
+		if !ok {
+			return errors.New("cannot cast received object to v1alpha1.EnvironmentMapping type")
```
Comment (piotrmiskiewicz) [inline components/remote-environment-broker/internal/mapping/controller.go]: add type information to the error message

File: components/remote-environment-broker/internal/mapping/controller.go [hunk 4]
```diff
@@ -157,47 +171,96 @@ func (c *Controller) processItem(key string) error {
...
+	if err := c.nsBrokerFacade.Create(envMapping.Namespace, c.systemNamesapce); err != nil {
+		return errors.Wrapf(err, "while creating namespaced broker in namespace [%s]", envMapping.Namespace)
 	}
-	err = c.applyNsAccLabel(reNs, label)
+
+	return nil
+}
+
+func (c *Controller) ensureNsBrokerNotRegistered(namespace string) error {
+	brokerExist, err := c.nsBrokerFacade.Exist(namespace)
```
Comment (piotrmiskiewicz) [inline components/remote-environment-broker/internal/mapping/controller.go]: why do we check if broker exists? Let's remove the check. The brokerFacade.Create skips error AlreadyExists.
Current implementation may be error prone - imagine service was not created (k8s API was not available for the service creation). The implementation of the Facade goes to the next operation - create Broker. At the end - broker is created, service - not. Error will be returned. The controller operation will be retried, but broker exists, so nothing will be done.
other solution - add checking if service exists in the Exists method

File: components/remote-environment-broker/internal/nsbroker/facade_test.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+func fixError() error {
+	return errors.New("some error")
+}
+
+func fixAlreadyExistError() error {
+	return k8s_errors.NewAlreadyExists(schema.GroupResource{}, "")
+}
+
+func fixNotFoundError() error {
```
Comment (piotrmiskiewicz) [inline components/remote-environment-broker/internal/nsbroker/facade_test.go]: minor: not used

File: components/remote-environment-broker/internal/mapping/controller.go (hunk 4 above)
Comment (aszecowka) [reply components/remote-environment-broker/internal/mapping/controller.go]: Exist now will check additionally if Service exist

**Pull request merged**